qiskit/circuit/quantumcircuit.py

```diff
@@ -499,6 +522,48 @@ def __eq__(self, other) -> bool:
             other, copy_operations=False
         )
 
+    def __copy__(self):
+        # Overridden to prevent copy.copy from using __{get,set}state__
+        # which currently unpacks _data into a Python list.
+        cls = self.__class__
+        cpy = cls.__new__(cls)
+        cpy.__dict__.update(self.__dict__)
+        return cpy
+
+    def __deepcopy__(self, memo=None):
+        # This is overridden in addition to __{get,set}state__
+        # to minimize memory pressure when we don't actually
+        # need to pickle (i.e. the typical deepcopy case).
+        cls = self.__class__
+        result = cls.__new__(cls)
+        for k in self.__dict__.keys() - {"_intern_context", "_data"}:
+            setattr(result, k, copy.deepcopy(self.__dict__[k], memo))
+
+        # Reuse the intern context to save space!
+        # TODO: there appears to be a concurrency issue when this
+        #   is done in a multi-threaded context (e.g. w/ primitives).
+        # result._intern_context = self._intern_context
+        result._intern_context = InternContext()
+
+        # Avoids pulling self._data into a Python list
+        # like we would when pickling.
+        result._data = result._new_data((copy.deepcopy(i, memo) for i in self._data))
+        return result
+
+    def __getstate__(self):
+        # Rust's CircuitData is not picklable, so we unpack it into a list first.
+        # InternContext is not picklable, so we leave it behind.
+        state = self.__dict__.copy()
+        state["_data"] = list(self._data)
+        del state["_intern_context"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+        self._intern_context = InternContext()
+        data = self._new_data(self._data)
+        self._data = data
+
```

Comment:
This ended up being a lot of work and adds a lot of Python-space complexity to stuff that's in practice called quite a lot.  I'm a bit concerned about that - it's really starting to feel like it would be better for us to properly define pickling for `InternContext` and `CircuitData`, since really we do need it.

Reply:
Done in 10cc624 and 0755ea5, with the exception of `__deepcopy__` which is still best implemented here due to missing support for `__deepcopy__` (technically you can override it in PyO3, but there's no PyO3 API for invoking `copy.deepcopy(memo)` on child objects).

In a future PR I'll put out shortly (to merge after this one), we speed up this implementation of `__deepcopy__` even further.